libweighted_tree/print_hierarchy.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2020 Lawrence Livermore National Security, LLC
+(c.f. AUTHORS, NOTICE.LLNS, COPYING)
+
+This file is part of the Flux resource manager framework.
+For details, see https://github.com/flux-framework.
+
+SPDX-License-Identifier: LGPL-3.0
+*/
+#include <iostream>
+#include <sqlite3.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <vector>
+#include <tuple>
+
+std::string hierarchy = "Bank|User|RawShares\n";
+std::string indent = "";
+char* messageError;
+sqlite3_stmt *res;
+sqlite3_stmt *select_stmt;
+
+int get_sub_banks(sqlite3* DB, const std::string& bank_name, const std::string& indent) {
+  int exit = 0;
+
+  // get shares of the current bank
+  std::string select_shares_stmt = "SELECT bank_table.shares "
+          "FROM bank_table "
+          "WHERE bank=?";
+  // compile SQL statement with bank name
+  exit = sqlite3_prepare_v2(DB, select_shares_stmt.c_str(), -1, &select_stmt, 0);
```

Comment:
Looking much better!  One additional change that could make things better:

One of the major benefits of using the `sqlite3_prepare_v2` function is that you only need to prepare a statement 1 time for the program's entire execution.  Right now, this statement gets prepared for every call of `get_sub_bank` (which is called many times recursively).  I would move the prepare statements out of this function and into `main`.  That will require a unique `sqlite3_stmt *` for each statement that is prepared.  The `sqlite3_bind_*` functions will be the ones that stay within this recursive function.

Also, for every corresponding call to `sqlite3_prepare_v2` there should be a call to the destructor [`sqlite3_finalize`](https://sqlite.org/c3ref/finalize.html).  Similar to how malloc/free and new/delete should be balanced.

Reply:
Ah, I see what you are saying. I think that makes sense.

From trying to convert it this morning, I found that I was able to `prepare` the first SQL statement (the one that gets the shares of the bank passed in), but not the other two (the ones that get all potential sub banks and associations). 

My guess is this has something to do with the value returned when `sqlite3_step()` returns something other than `SQLITE_OK` (according to their documentation, the resulting value returned is undefined), which results in a `Segmentation fault`. I will try and work with it some more to see if I can get it working.

Reply:
Well, I guess after enough `cout` statements, I have found the culprit. 😆

Before each recursive call, I think the SQL statements must have their bindings cleared and reset so that the new bindings can be placed in:

```c++
sqlite3_clear_bindings(stmt);
sqlite3_reset(stmt);
```

Now I can `prepare` each SQL statement just once in `main()`, and `bind` the parameters when `get_sub_banks()` gets called. I will squash and push those changes up now.